### PR TITLE
add `initial_value` to `aie.buffer`

### PR DESCRIPTION
**Pull request by makslevental**
This PR adds a way to initialize `aie.buffer`:

```mlir
%buf33 = aie.buffer(%t33) : memref<2x2xi32> = dense<[[0, 1], [2, 3]]>
%buf42 = aie.buffer(%t42) : memref<2x3xi32> = dense<[[0, 1, 2], [3, 4, 5]]>
%buf44 = aie.buffer(%t44) : memref<3x2xi32> = dense<[[0, 1], [2, 3], [4, 5]]>
```

which gets lowered to (through `--aie-standard-lowering`):

```mlir
memref.global "public" @buf44 : memref<3x2xi32> = dense<[[0, 1], [2, 3], [4, 5]]>
memref.global "public" @buf42 : memref<2x3xi32> = dense<[[0, 1, 2], [3, 4, 5]]>
memref.global "public" @buf33 : memref<2x2xi32> = dense<[[0, 1], [2, 3]]>
```

In order to effectively make use of this new feature you need to emit the BCF linker script on a core-by-core basis and compile the core code on a core-by-core basis. The reason for this is because otherwise the globals will simply be duplicated across all core elf files. With the current arrangment you get

```llvm
;;;;; core_1_3.ll
; buffer owned but not inited by core tile (1, 3)
@buffer_1_3_result = external global [32 x i32]
; buffer owned and inited by core tile (1, 3)
@buffer_1_3_inited_access_elsewhere = global [32 x i32] [i32 7, i32 6, i32 3, ...]
```

```llvm
;;;;; core_2_3.ll
; buffer owned but not inited by core tile (2, 3)
@buffer_weight_2_3_result = external global [32 x i32]
; external buffers from neigbhoring tiles 
@buffer_weight_2_2 = external global [32 x i32]
@buffer_weight_2_4 = external global [32 x i32]
@buffer_1_3_result = external global [32 x i32]
; external and inited buffer on core tile (1, 3)
@buffer_1_3_inited_access_elsewhere = external global [32 x i32]
```

**Note**, the symbol `buffer_1_3_inited_access_elsewhere` needs to point to the correct places in both the address space of `core (1,3)` and `core (2, 3)`, i.e., `buffer_1_3_inited_access_elsewhere` needs to point to some address assigned by `aie-assign-addresses` (which corresponds to a fixed address in `core (1,3)` and to some known offset in `core (2, 3)` corresponding to `core (1, 3)`'s mmaped address space) _and_ wherever the data segment of the object file deigns to place it. This magic trick is accomplished by using the `_overlay` keyword in the BCF linker script (section 4.8 in the Bridge manual).

My python E2E flow has been updated to demonstrate how to use the feature; roughly:

```python
for col, row, _ in generate_cores_list(module):
    # generate bcf for col, row
    core_bcf = generate_bcf(core_mod.operation, col, row)
    # lower to LLVM but only the aie.core i.e. func.func that corresponds to col, row
    core_lowered_to_llvm_dialect = run_pipeline(core_mod, AIE_LOWER_TO_LLVM(col, row))
    core_input_ll = translate_mlir_to_llvmir(core_lowered_to_llvm_dialect.operation)
    # compile and output a named object file (core_{col}_{row}.o)
    chess_compile(
        link_with_chess_intrinsic_wrapper(core_input_ll),
        output_filename=f"core_{col}_{row}",
    )
    # link everything together using the named object file
    make_core_elf(core_bcf, input_filename=f"core_{col}_{row}")
```

### CODE REVIEW

File: lib/Dialect/AIE/IR/AIEDialect.cpp [hunk 1]
```diff
@@ -358,10 +359,12 @@ void AIEDialect::initialize() {
   addAttributes<
 #define GET_ATTRDEF_LIST
 #include "aie/Dialect/AIE/IR/AIEAttrs.cpp.inc"
+
```
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:362]: odd whitespace change?

File: test/python/tile_array.py [hunk 2]
```diff
@@ -253,9 +253,9 @@ def ipu():
         print(find_neighbors(tiles[2, 2].df))
 
         assert tiles[1:3, 1:3].neighbors.shape == (2, 2)
-        # CHECK: tile(col=1, row=1) Neighbors(north=%tile_1_2 = aie.tile(1, 2), west=None, south=None)
-        # CHECK: tile(col=1, row=2) Neighbors(north=%tile_1_3 = aie.tile(1, 3), west=%tile_0_2 = aie.tile(0, 2), south=%tile_1_1 = aie.tile(1, 1))
-        # CHECK: tile(col=2, row=1) Neighbors(north=%tile_2_2 = aie.tile(2, 2), west=%tile_1_1 = aie.tile(1, 1), south=None)
-        # CHECK: tile(col=2, row=2) Neighbors(north=%tile_2_3 = aie.tile(2, 3), west=%tile_1_2 = aie.tile(1, 2), south=%tile_2_1 = aie.tile(2, 1))
+        # CHECK: tile(col=1, row=1) : Neighbors(north=%tile_1_2 = aie.tile(1, 2), west=None, south=None)
```
Comment (stephenneuendorffer) [inline test/python/tile_array.py:256]: unrelated change?

**Review by stephenneuendorffer — APPROVED** [review]

File: test/python/tile_array.py (hunk 2 above)
Comment (makslevental) [reply test/python/tile_array.py:256]: These are effectively tests (and basically my whole goal) for the feature: grab neighboring tiles, grab buffers in those tiles. Problem became (took the whole day), that the BCF linker and clang don't cooperate when it comes to data in `rodata`. Hence "add `initial_value` to `aie.buffer`".

**Comment by keryell** [comment]
Very interesting PR description!
It would be so nice to have this kind of information stored in some design document or at least comments instead of being lost for ever...
